File: desafio4/kotlin/src/main/kotlin/huffman/BitOutputStream.kt
package huffman

import java.io.OutputStream

/**
 * Output bits on a stream
 * Created by ediaz on 5/21/17.
 */
class BitOutputStream(val output : OutputStream) {

    var buffer : Int = 0
    var bitsInBuffer : Int = 0


    private fun writeBit(bit: Int) {
        if (bit != 0 && bit != 1)
            throw IllegalArgumentException("Argument must be '0' or '1'. Received: "+bit)

        buffer = (buffer shl 1) or bit
        bitsInBuffer++
        if (bitsInBuffer == 8) {
            clearBuffer()
        }
    }

    private fun writeByte(byte:Int) {
        assert(byte >= 0 && byte <= 256)
        for (i in 0..8){
            val bit = ((byte ushr (8 - i -1)) and 1)
            writeBit(bit)
        }
    }

    private fun clearBuffer() {
        output.write(buffer)
        buffer = 0
        bitsInBuffer = 0
    }


    fun flush() {
        clearBuffer()
        output.flush()
    }

    fun close() {
        flush()
        output.close()
    }

    fun write(b: Boolean) {
        writeBit(if (b) 1 else 0)
    }

    fun write(b: Byte) {
        writeByte(b.toInt() and 0xFF)
    }


    fun write(i: Int) {
        writeByte((i ushr 24) and 0xFF)
        writeByte((i ushr 16) and 0xFF)
        writeByte((i ushr  8) and 0xFF)
        writeByte((i ushr  0) and 0xFF)
    }

    fun write(c: Char) {
        writeByte(c.toInt() and 0xFF)
    }

}

File: desafio4/scala1/src/main/scala/bitio/BitOutputStream.scala
package bitio

import java.io.OutputStream

/**
  * Output bits on a stream
  * Created by ediaz on 7/15/17.
  */
class BitOutputStream(val output: OutputStream) extends BitStream(output) {

  private[this] def writeBit(bit: Int) : Unit = {
    if (bit != 0 && bit != 1) {
      throw new IllegalArgumentException("Argument must be 0 or 1. Received: "+bit)
    }

    buffer = (buffer << 1) | bit
    bitsInBuffer += 1
    if (bitsInBuffer == 8) {
      clearBuffer()
    }
  }

  private[this] def clearBuffer(): Unit = {
    if (bitsInBuffer == 0) {
      return
    }
    if (bitsInBuffer > 0) {
      buffer = buffer << (8 - bitsInBuffer)
    }
    output.write(buffer)
    buffer = 0
    bitsInBuffer = 0
  }

  private[this] def writeByte(byte: Int) : Unit = {
    if (byte < 0 || byte >= 256)
      throw new IllegalArgumentException("Argument must be in range 0..255. Received: "+byte)

    (0 to 7).foreach { i =>
      val bit = (byte >>> (8 - i - 1)) & 1
      writeBit(bit)
    }
  }

  def flush() : Unit = {
    clearBuffer()
    output.flush()
  }

  override def close(): Unit = {
    flush()
    super.close()
  }

  def write(b: Boolean) : Unit = {
    writeBit(if (b) 1 else 0)
  }

  def write(i: Int): Unit = {
    writeByte((i >>> 24) & 0xFF)
    writeByte((i >>> 16) & 0xFF)
    writeByte((i >>>  8) & 0xFF)
    writeByte((i >>>  0) & 0xFF)
  }

  def write(c: Char) : Unit = {
    writeByte(c & 0xFF)
  }
}
